File: problems/CSES/Sorting_and_Searching/nested_ranges_count/nested_ranges_count.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;

struct P {
    ll x;
    ll y;
    int index;
};

struct Fenwick {
    int n;
    vector<int> bit;

    Fenwick(int num) : n(num), bit(num + 1, 0) {}

    void add(int i, int value) {
        while (i <= n) {
            bit[i] += value;
            i += i & -i;
        }
    }

    int prefix_sum(int i) {
        int sum_val = 0;

        while (i > 0) {
            sum_val += bit[i];
            i -= i & -i;
        }

        return sum_val;
    }
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    cin >> n;
        
    vector<P> ranges;
    ranges.reserve(n);

    vector<ll> all_y;
    all_y.reserve(n);

    for (int i = 0; i < n; i++) {
        ll x, y;
        cin >> x >> y;
        ranges.push_back({x, y, i});
        all_y.push_back(y);
    }
        
    sort(ranges.begin(), ranges.end(), [](const P& a, const P& b) {
        if (a.x != b.x) return a.x < b.x;
        return a.y > b.y;
    });

    sort(all_y.begin(), all_y.end());
    all_y.erase(unique(all_y.begin(), all_y.end()), all_y.end());
    auto get_comp_y = [&](ll y) {
        return lower_bound(all_y.begin(), all_y.end(), y) - all_y.begin() + 1;
    };

    int max_y_idx = all_y.size();
    vector<int> contains(n, 0);
    vector<int> contained_by(n, 0);

    Fenwick bit1(max_y_idx);
    for (int i = n - 1; i >= 0; i--) {
        int pos = get_comp_y(ranges[i].y);
        contains[ranges[i].index] = bit1.prefix_sum(pos);
        bit1.add(pos, 1);
    }

    Fenwick bit2(max_y_idx);
    int processed = 0;
    for (int i = 0; i < n; i++) {
        int pos = get_comp_y(ranges[i].y);
        contained_by[ranges[i].index] = processed - bit2.prefix_sum(pos - 1);
        bit2.add(pos, 1);
        processed++;
    }

    for (int i = 0; i < n; i++) cout << contains[i] << (i == n - 1 ? "" : " ");
    cout << "\n";

    for (int i = 0; i < n; i++) cout << contained_by[i] << (i == n - 1 ? "" : " ");
    cout << "\n";

    return 0;
}

File: problems/CSES/Sorting_and_Searching/nested_ranges_count/nested_ranges_count.py
import sys
# sys.setrecursionlimit(200000)

input = sys.stdin.readline

# Russian:
"""
Этот код полностью логически правильный, но изза особености Python (в особенности его скорости)
это решение не пройдет, и как мне кажеться, найти решение на Python которое пройдет тесты будет
весьма проблематично.
"""

# English:
"""
This code is entirely logically correct, but due to a peculiarity of Python (particularly its speed),
this solution will not pass the tests, and I think it will be quite difficult to find a Python solution 
that will pass the tests.
"""

class Fenwick:
    def __init__(self, n):
        self.n = n
        self.bit = [0] * (n + 1)

    def add(self, i, value):
        while i <= self.n:
            self.bit[i] += value
            i += i & -i

    def prefix_sum(self, i):
        res = 0

        while i > 0:
            res += self.bit[i]
            i -= i & -i

        return res

def main():
    n = int(input())
    ranges = []
    for i in range(n):
        x, y = map(int, input().split())
        ranges.append((x, y, i))

    ranges.sort(key=lambda x: x[1], reverse=True)
    ranges.sort(key=lambda x: x[0])

    ys = sorted(list(set(r[1] for r in ranges)))
    comp = {y: i + 1 for i, y in enumerate(ys)}

    contains = [0] * n
    contained_by = [0] * n

    bit = Fenwick(len(ys))

    for x, y, idx in reversed(ranges):
        pos = comp[y]
        contains[idx] = bit.prefix_sum(pos)
        bit.add(pos, 1)

    bit = Fenwick(len(ys))
    processed = 0

    for x, y, idx in ranges:
        pos = comp[y]
        contained_by[idx] = processed - bit.prefix_sum(pos - 1)
        bit.add(pos, 1)
        processed += 1

    print(*contains)
    print(*contained_by)

main()
